경북대 BE_유우석 - Step1 기능 구현

milk-stone opened the pull request:
Step1 기능을 구현하였습니다. <br>

구현한 기능은 README.md 파일에 작성하였습니다.

### 질문

- 인터넷을 찾아보니, 금지어를 "카카오" 처럼 설정하는 경우 **사용자 정의 어노테이션**을 만들어 사용하는 경우가 있던데, 이런 방법으로 작성하는 것이 더 좋을까요? 여러 금지어 설정할 수 있다는 점에서 코드 재사용성이 좋아질 것 같다는 생각이 들었으나 실제로도 이런 코드를 추천하시는지 궁금해서 여쭤봅니다!

milk-stone committed:
docs(readme): Step 1에 진행할 내용 정리

milk-stone committed:
chore(build.gradle): Validation 적용을 위한 의존성 추가

- 아래와 같이 추가하면 build가 되지 않음.
implementation 'spring-boot-starter-validation'

- 앞에 org.springframework.boot: 를 추가하니 정상적으로 빌드됨.
implementation 'org.springframework.boot:spring-boot-starter-validation'

milk-stone committed:
feat: 상품 생성, 수정 DTO에 Validation 코드 추가

상품 이름
- @NotBlank : 필수 항목 선언
- @Size : 상품 이름 최대 길이 제한
- @Pattern : 특수문자 여부와 *카카오*가 들어있는지 여부 체크

상품 가격
- @NotNull : 필수 항목 선언
- @Min : 음수가 들어오지 않도록 설정
- @Max : int 최댓값을 넘지 않도록 수정 (DB에 저장할 때 에러가 나지 않게 하기 위함)

이미지 링크
- @NotNull : 필수 항목 선언

ProductController.java
- @Valid 태그를 추가하여 유효성 검증이 작동하게 함.

milk-stone committed:
feat(exception): 전역 예외 핸들러를 설정하여 Validation 시 발생하는 MethodArgumentNotValidException에 대한 에러 메세지 가공

milk-stone committed:
feat(products.html): 관리자 화면에서 상품 추가, 수정 시 유효성을 검증하고, 위반한 사항에 대한 에러 메세지를 표시하도록 코드 작성

milk-stone committed:
docs(readme): 1단계 진행 상황 체크

Rok93 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<ErrorResponse> handleValidationExceptions(MethodArgumentNotValidException e) {
+
+        // 1. 예외 객체에서 모든 필드 에러 정보를 가져오기
```
학습 열심히 해주신게 눈에 보이네요. 👍👍 

다만 실제 프로덕트라고 생각했을 때, 이런 주석들은 조금 불필요하다고 생각해요. 👀
물론 현업에서도 아예 주석을 사용하지 않는 것은 아니에요. 
다만 가급적 주석을 먼저 쓰기보다는 코드를 통해서 최대한 표현하되 코드로는 표현하기 힘든 어려운 것들을 주석으로 남기는 편이에요. 

지금 남겨주신 주석들은 꼭 필요한거라고 생각하시나요? 😃

milk-stone replied:
피드백 감사합니다! 

평소 주석을 작성할 때, 코드를 잘 모르는 사람도 주석을 통해 이해할 수 있게 하는걸 목표로 했던 것 같은데, 실제 프로덕트라고 생각해보니 저도 이런 주석은 불필요하다고 느껴지네요! (객체 지향 체조 원칙을 지켜 코드를 작성한다면, 충분히 많은 사람이 이해할 수 있을 것 같습니다.)

제가 적어놓은 이 주석들은 코드 구조로 파악이 가능한 내용들이니, 피드백 반영하여 주석 삭제 하여 다시 제출해보겠습니다! 피드백 감사합니다! 😄

Rok93 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+                .collect(Collectors.toList());
+
+        // 4. 최종 에러 응답 객체를 생성.
+        ErrorResponse errorResponse = new ErrorResponse("입력 값에 대한 유효성 검사에 실패했습니다.", fieldErrors);
```
에러응답을 위한 ErrorResponse를 만드셨군요?  👍👍 

첨언을 드리자면 ErrorResponse, CustomFieldError 클래스는 가급적 별도 클래스로 분리해보면 어떨까요? 😃

milk-stone replied:
앗 그렇네요! 지금은 핸들링 하는 예외 상황이 하나라 이정도 코드로 가독성이 괜찮겠지만, 나중에 핸들링하는 에러가 많아지면 ErrorResponse record의 위치를 찾아 수정하는게.. 힘들어질 것 같네요!

클래스 분리하여 제출하겠습니다!

Rok93 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,35 @@
+        // 5. HTTP 400 (Bad Request) 상태 코드와 함께 에러 응답을 반환.
+        return new ResponseEntity<>(errorResponse, HttpStatus.BAD_REQUEST);
+    }
+
```
만약 `MethodArgumentNotValidException` 이외의 다른  예외가 발생하게되는 경우에는 어떻게 될까요? 🤔

가능하면 예기치못한 에러가 발생했을 때에는 서버 에러 응답인 500 에러 응답을 내려주도록해주면 조금 더 깔끔한 일관된 응답을 해줄 수 있지 않을까요? 😁

milk-stone replied:
예외 클래스에도 계층 구조가 있으니까 처리하지 못한 클래스에 대해서도 에러메세지를 가공해서 주면 View 계층에서 이해하기 쉬워질 것 같네요!

이것 역시 반영해보겠습니다!

Rok93 commented on src/main/java/gift/product/dto/ProductRequest.java:
```diff
@@ -1,4 +1,25 @@
+
+        @NotNull(message = "가격은 비워둘 수 없습니다.")
+        @Min(value = 0, message = "가격은 0원 이상이어야 합니다.")
+        @Max(value = 2147483647, message = "가격은 2147483647원 이하이어야 합니다.")
```
음... 가격의 최대치 제한은 요구사항에 있는 내용일까요? 👀  
조금 뜬금없는 숫자라고 생각이돼서요. 🤔

milk-stone replied:
요구사항에는 가격에 대한 제한은 없었는데, 실제로 제가 만든 관리자 페이지로 테스트를 해보다가 INT 최대값을 초과하는 경우 INSERT 문에서 오류가 나더라고요!
(DB 테이블 생성할 때 `price INT NOT NULL` 이렇게 설정했었습니다.)

그래서 유효성 검증을 하고자 이렇게 작성했었는데, 이 숫자가 뜬금 없이 느껴지는 것 같기도 합니다. 2억원 정도로 수정해보겠습니다!

Rok93 commented on src/main/java/gift/product/dto/ProductUpdateRequest.java:
```diff
@@ -1,4 +1,25 @@
+        @NotBlank(message = "상품 이름은 비워둘 수 없습니다.")
+        @Size(max = 15, message = "상품 이름은 공백을 포함하여 최대 15자까지 입력할 수 있습니다.")
+        @Pattern(
+                regexp = "^[a-zA-Z0-9가-힣\\s\\(\\)\\[\\]\\+\\-\\&\\/_]*$",
```
사실 생각해보면 이런 `상품 이름 정책`은 **도메인 정책**으로 볼 수 있을 것 같아요. 
가능하다면 `Product` 클래스 내부에 이런 비즈니스 로직을 보장해주면 좋지 않을까요? 😃

Product 객체가 생성되는 시점에 이런 상품명에 대한 유효성 검증을 진행해보도록 수정해보면 어떨까요? 🤗

milk-stone replied:
Product 내부에서 도메인 관련 유효성 검증을 진행하는 방향으로 수정해보겠습니다!

한 번도 생각해보지 못했던 부분인데, 짚어 알려주셔서 감사합니다!

Rok93 commented on src/main/java/gift/product/dto/ProductRequest.java:
```diff
@@ -1,4 +1,25 @@
+                message = "상품 이름에 허용되지 않는 특수문자가 포함되어 있습니다. ( ), [ ], +, -, &, /, _ 만 사용 가능합니다."
+        )
+        @Pattern(
+                regexp = "^(?!.*카카오).*$",
```
> 인터넷을 찾아보니, 금지어를 "카카오" 처럼 설정하는 경우 사용자 정의 어노테이션을 만들어 사용하는 경우가 있던데, 이런 방법으로 작성하는 것이 더 좋을까요? 여러 금지어 설정할 수 있다는 점에서 코드 재사용성이 좋아질 것 같다는 생각이 들었으나 실제로도 이런 코드를 추천하시는지 궁금해서 여쭤봅니다!

음... 크게 고민해보지는 않았는데 어디까지 고려하냐에따라 조금 다를 것 같아요. 😃
"카카오"와 같은 워딩을 어떻게 생각해보면 `금지어` 같은 개념으로 생각해볼 수도 있을 것 같은데요. 
금지어는 추후에 변경될 수 있고 하나가 아니라 여러개로 추가될 수도 있다고 생각해요. 
하지만 지금처럼 `@Pattern` 방식(
[ConstraintValidator 방식](https://tecoble.techcourse.co.kr/post/2021-06-21-custom-annotation/)도 동일)은 결국 금지어가 변경되거나 혹은 추가가되는 경우 코드를 수정하여 재배포를해야하게될 것 같아요. 

이런 면에서는 말씀하신 방식들이 그렇게 효과적인지는 잘 모르겠어요. 🤔 

차라리 상품명 검증 역할을하는 객체를 별도로 만들어서 처리를 하는 것이 조금 더 유연한 대응이 되지 않을까 싶어요. 👀

Rok93 replied:
또 하나 조금 고민되는 포인트가하나 있는데요.
만약 상품명에 '카카오'가 들어가게됐을 경우 프론트에서 다른 처리 (예를들어 MD에게 문의하는 프로세스로 연결)를 하게된다면 어떻게될까요? 🤔

지금의 방식은 조금 문제가 있지 않을까요? 👀

어떤 문제점이있는지 한번 고민해보신 뒤 코멘트로 남겨주시겠어요?
추가적으로 이 문제점을 해결하기위한 방법도 같이 고민해보고 떠오르는 방법이 있다면 같이 남겨주시면 좋을 것 같아요. 🤗

milk-stone replied:
@Pattern 으로 처리를 하던, 커스텀 어노테이션으로 처리를 하던 입력 검증 차원에서 막아버리게 되네요! 

| 만약 상품명에 '카카오'가 들어가게됐을 경우 프론트에서 다른 처리 (예를들어 MD에게 문의하는 프로세스로 연결)를 하게된다면 어떻게될까요? 🤔
말씀해주신 것처럼, 이런 로직이 추가되어야 한다면 Service 단에서 입력 값을 분석하여 처리를 하는 것이 조금 더 용이할 것 같습니다!

@Valid 태그를 이용해 검증하는 것은, 입력 값이 null 이라거나, 이상한 문자가 들어왔다거나 등 **입력 데이터를 분석하지 않아도 문제**인 입력값에 대해 검증하도록 하고,

입력값이 정상 범주인 경우에는, 그 값에 따라 추가 로직을 수행할 수 있는 가능성을 열어두는 것이 좋은 것 같습니다! ('카카오'가 들어있을 경우 MD와 연결 등..)

Rok93 commented on src/main/java/gift/product/dto/ProductRequest.java:
```diff
@@ -1,4 +1,25 @@
+        String name,
+
+        @NotNull(message = "가격은 비워둘 수 없습니다.")
+        @Min(value = 0, message = "가격은 0원 이상이어야 합니다.")
```
상품 등록시 0원도 허용이 되는걸까요? 🤔

milk-stone replied:
무료인 상품이 있을 수는 없겠네요... 조금 더 이 기능이 무엇을 위해 존재하는지 생각하고 코드를 짜는 습관을 들여야겠습니다! 피드백 감사합니다!

Rok93 replied:
이랬다 저랬다 말해서 죄송해요.  👀

사실 저는 무료인 상품도 있을 수 있겠다...! 라고 생각했어요. (이벤트성) 
다만 이런 경우에는 너무 많은 사람들이 구매를하면 안되니 선착순 N명 이라거나 등의 기능으로 제한을 해줘야하지 않을까 싶은데 지금은 정해진 미션에 맞춰 스펙을 구현해야하니 이런 관점도 있겠다...! 정도로 생각해주시면 좋을 것 같아요. 

반영해주신 방식이 지금에는 조금 더 적절하다고 생각해요. 😉

Rok93 commented on src/main/resources/templates/products.html:
```diff
@@ -262,6 +281,29 @@ <h3 id="viewProductName"></h3>
+            $(`#${formId} .form-control`).removeClass('is-invalid');
+        }
+
+        function displayErrors(formId, errors) {
```
<img width="1474" alt="image" src="https://github.com/user-attachments/assets/da65c2d5-b625-4519-87d7-1ae2e2039634" />

오오... 에러가 발생한 원인들을 폼으로도 동적으로 잘 노출해주고있군요? 👍🤩

milk-stone replied:
감사합니다!!

Rok93 commented on src/main/resources/templates/products.html:
```diff
@@ -262,6 +281,29 @@ <h3 id="viewProductName"></h3>
+        }
+
+        // 모달이 닫힐 때 에러 메시지 초기화
+        $('#addProductModal, #editProductModal').on('hidden.bs.modal', function () {
```
ES6에 나온 [화살표 함수](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Functions/Arrow_functions)를 활용해도 좋을 것 같아요. 😃 

```suggestion
        $('#addProductModal, #editProductModal').on('hidden.bs.modal', () => {
```

milk-stone replied:
감사합니다! 추천해주신 화살표 함수 관련해서도 학습해보도록 하겠습니다!

제 프론트엔드 지식 수준이 웹 코드 구조를 이해하는 정도에 그치고 있어, 사실 화살표 함수? 이런게 익숙하지 않긴 하지만, 좋은 개발자가 되기 위해서는 프론트 백 나눌 것 없이 다 알아야한다고 생각하기에 **꼭 학습 하겠습니다!** (현재는 LLM을 이용해 바이브 코딩을 진행하고 있습니다 !!)

Rok93 replied:
꼭 모든 것을 다 알고있을 필요는 없어요. 
저는 개인적으로 js 공부에도 관심이 있어 한 때, 했던 경험(?)이 있어서 종종 도움이될 떄가 있어요. 
말씀하신 것처럼 요즘에는 AI 툴을 잘 활용하는 것도 하나의 방법이 될 수 있다고 생각해요. 😉

Rok93 reviewed (changes requested):
우석님 1단계 구현 잘해주셨습니다. 👍💯
몇몇 코멘트 남겨두었는데 확인해서 반영 부탁드릴게요. 🤗
질문에 대한 제 생각도 남겨두었으니 같이 확인해주세요!

milk-stone committed:
refactor: CustomFieldError, ErrorResponse record 클래스 분리, 예기치 못한 에러 처리를 위한 Exception.class 핸들러 함수 작성

- 불 필요한 주석 삭제
- 내부 코드 상에서 발생하는 에러 메세지를 담기 위한 ProblemDetail 클래스를 임포트함.

milk-stone committed:
refactor: 3가지 변경 사항 적용

1. IllegalArgumentException 핸들러 함수 추가

2. ProblemDetail 대신 에러 메세지만 담는 SingleErrorResponse 레코드 클래스를 작성하여 사용

3. Logger를 추가하여 spring console에 로그가 적히도록 설정

milk-stone committed:
fix: 기본 데이터를 생성해주는 data.sql가 상품 이름 정책을 위반하는 데이터를 넣던 현상 수정

- 상품 이름이 15글자를 초과한 행 수정

milk-stone committed:
refactor: 2가지 변경 사항

1. 가격에 대한 변경 사항
- 최솟값을 1원으로 설정 (무료인 상품은 없으므로..)
- 최댓값을 2억원으로 설정

2. 상품 이름에 대한 변경 사항
- 빈 값이 들어오지 않도록 하는 어노테이션만 유지함.
- Product 객체 생성 시점에 유효성을 검증할 수 있도록 Product 클래스 내부에 검증 메서드 작성

milk-stone committed:
feat(products.html): 관리자 페이지에서 상품 이름에 대한 도메인 정책 위반 내역을 alert 할 수 있도록 handleAjaxError 함수 작성

Rok93 commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,41 @@
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<SingleErrorResponse> handleIllegalArgumentExceptions(IllegalArgumentException e) {
+        log.warn("비즈니스 로직 유효성 검사 실패: {}", e.getMessage());
```
오... 경고 로그와 에러 레벨의 로그를 사용하셨는데요. 
경고 레벨을 이렇게 설정하신 이유가 궁금합니다! 😃

Rok93 commented on src/main/java/gift/product/Product.java:
```diff
@@ -1,15 +1,21 @@
     private String imageUrl;
 
+    private static final int MAX_NAME_LENGTH = 15;
+    private static final Pattern ALLOWED_NAME_PATTERN = Pattern.compile("^[a-zA-Z0-9가-힣\\s\\(\\)\\[\\]\\+\\-\\&\\/_]*$");
```
값비싼 연산인 Pattern을 상수로 두셨군요? 😃 
이미 알고 계실 수도 있지만 아래의 참고자료도 한번 읽어보시면 도움이 되실 것 같아요. 😉

[[아이템6] 불필요한 객체 생성을 피하라](https://lovethefeel.tistory.com/141)

Rok93 commented on src/main/java/gift/product/Product.java:
```diff
@@ -1,15 +1,21 @@
     }
 
     public Product(String name, Long price, String imageUrl) {
+        validateName(name);
```
이제는 어디서든 Product 생성에 대해서 일관적인 규칙을 적용할 수 있겠네요. 👍💯

Rok93 commented on src/main/java/gift/product/dto/ProductRequest.java:
```diff
@@ -1,4 +1,16 @@
+
+        @NotNull(message = "가격은 비워둘 수 없습니다.")
+        @Min(value = 1, message = "가격은 1원 이상이어야 합니다.")
+        @Max(value = 200000000, message = "가격은 2억원 이하이어야 합니다.")
```
사소할 수 있지만 3자릿수마다 `_`를 넣어준다면 조금 더 한눈에 숫자를 파악할 수 있을 것 같아요. 😃 

```suggestion
        @Max(value = 200_000_000, message = "가격은 2억원 이하이어야 합니다.")
```

Rok93 reviewed (approved):
우석님 1단계 피드백 반영 잘해주셨어요. 
1단계는 충분히 잘 구현해주신 것 같아서 이만 머지하도록하겠습니다. 🤗
다만 다음 단계 제출일자까지 조금 여유가있는만큼 이번에는 테스트 코드 작성에 도전해보시면 어떨까싶어요! 😃 
미션 진행하시면서 어려운 부분 있으시면 언제든지 DM 주세요. 
2단계 기대하고있을께요! 😉
